File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/matchers/ProxyResponseAssertions.java
package uk.nhs.nhsx.matchers;

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeDiagnosingMatcher;
import uk.nhs.nhsx.core.exceptions.HttpStatusCode;

public class ProxyResponseAssertions {

    public static Matcher<APIGatewayProxyResponseEvent> hasStatus(HttpStatusCode code) {
        return new TypeSafeDiagnosingMatcher<APIGatewayProxyResponseEvent>() {
            @Override
            protected boolean matchesSafely(APIGatewayProxyResponseEvent event, Description description) {
                Integer actual = event.getStatusCode();
                description.appendText("status was ").appendValue(actual);
                return actual != null && actual.equals(code.code);
            }

            @Override
            public void describeTo(Description description) {
                description.appendText("code ").appendValue(code);
            }
        };
    }

    public static Matcher<APIGatewayProxyResponseEvent> hasHeader(String name, Matcher<String> value) {
        return new TypeSafeDiagnosingMatcher<APIGatewayProxyResponseEvent>() {
            @Override
            protected boolean matchesSafely(APIGatewayProxyResponseEvent event, Description mismatch) {
                String actual = event.getHeaders().get(name);
                if (!value.matches(actual)) {
                    mismatch.appendText("body").appendText(" ");
                    value.describeMismatch(actual, mismatch);
                    return false;
                } else {
                    return true;
                }
            }

            @Override
            public void describeTo(Description description) {
                description.appendText("header ").appendValue(name).appendText(" ").appendDescriptionOf(value);
            }
        };
    }

    public static Matcher<APIGatewayProxyResponseEvent> hasBody(Matcher<String> body) {
        return new TypeSafeDiagnosingMatcher<APIGatewayProxyResponseEvent>() {
            @Override
            protected boolean matchesSafely(APIGatewayProxyResponseEvent event, Description mismatch) {
                String actual = event.getBody();
                if (!body.matches(actual)) {
                    mismatch.appendText("body").appendText(" ");
                    body.describeMismatch(actual, mismatch);
                    return false;
                } else {
                    return true;
                }
            }

            @Override
            public void describeTo(Description description) {
                description.appendText("body ").appendDescriptionOf(body);
            }
        };
    }
}

File: src/aws/lambdas/incremental_distribution/infra/src/test/kotlin/uk/nhs/nhsx/testhelper/matchers/ProxyResponseAssertions.kt
package uk.nhs.nhsx.testhelper.matchers

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.TypeSafeDiagnosingMatcher
import uk.nhs.nhsx.core.exceptions.HttpStatusCode

object ProxyResponseAssertions {

    fun hasStatus(code: HttpStatusCode): Matcher<APIGatewayProxyResponseEvent> {
        return object : TypeSafeDiagnosingMatcher<APIGatewayProxyResponseEvent>() {
            override fun matchesSafely(event: APIGatewayProxyResponseEvent, description: Description): Boolean {
                val actual = event.statusCode
                description.appendText("status was ").appendValue(actual)
                return actual != null && actual == code.code
            }

            override fun describeTo(description: Description) {
                description.appendText("code ").appendValue(code)
            }
        }
    }

    fun hasHeader(name: String, value: Matcher<String>): Matcher<APIGatewayProxyResponseEvent> {
        return object : TypeSafeDiagnosingMatcher<APIGatewayProxyResponseEvent>() {
            override fun matchesSafely(event: APIGatewayProxyResponseEvent, mismatch: Description): Boolean {
                val actual = event.headers[name]
                return if (!value.matches(actual)) {
                    mismatch.appendText("body").appendText(" ")
                    value.describeMismatch(actual, mismatch)
                    false
                } else {
                    true
                }
            }

            override fun describeTo(description: Description) {
                description.appendText("header ").appendValue(name).appendText(" ").appendDescriptionOf(value)
            }
        }
    }

    fun hasHeader(name: String): Matcher<APIGatewayProxyResponseEvent> {
        return object : TypeSafeDiagnosingMatcher<APIGatewayProxyResponseEvent>() {
            override fun matchesSafely(event: APIGatewayProxyResponseEvent, mismatch: Description): Boolean =
                event.headers[name] != null

            override fun describeTo(description: Description) {
                description.appendText("header ").appendValue(name)
            }
        }
    }

    fun hasBody(body: Matcher<String>): Matcher<APIGatewayProxyResponseEvent> {
        return object : TypeSafeDiagnosingMatcher<APIGatewayProxyResponseEvent>() {
            override fun matchesSafely(event: APIGatewayProxyResponseEvent, mismatch: Description): Boolean {
                val actual = event.body
                return if (!body.matches(actual)) {
                    mismatch.appendText("body").appendText(" ")
                    body.describeMismatch(actual, mismatch)
                    false
                } else {
                    true
                }
            }

            override fun describeTo(description: Description) {
                description.appendText("body ").appendDescriptionOf(body)
            }
        }
    }
}
